server/src/main/java/io/druid/server/coordinator/cost/SegmentCostCache.java

```diff
@@ -0,0 +1,379 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.coordinator.cost;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Ordering;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.guava.Comparators;
+import io.druid.server.coordinator.CostBalancerStrategy;
+import io.druid.timeline.DataSegment;
+import org.apache.commons.math3.util.FastMath;
+import org.joda.time.Interval;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class SegmentCostCache
+{
+  private static final double HALF_LIFE = 1.0; // cost function half-life in hours
+  private static final double LAMBDA = Math.log(2) / HALF_LIFE;
+  private static final double MILLIS_FACTOR = TimeUnit.DAYS.toMillis(1) / LAMBDA;
+  private static final long LIFE_THRESHOLD = TimeUnit.DAYS.toMillis(30);
+  private static final long BUCKET_INTERVAL = TimeUnit.DAYS.toMillis(30);
+
+  private static final Comparator<DataSegment> SEGMENT_INTERVAL_COMPARATOR = (d1, d2) ->
+      Comparators.intervalsByStartThenEnd().compare(
+          d1.getInterval(),
+          d2.getInterval()
+      );
+
+  private static final Comparator<Bucket> BUCKET_INTERVAL_COMPARATOR = (b1, b2) ->
+      Comparators.intervalsByStartThenEnd().compare(
+          b1.getInterval(),
+          b2.getInterval()
+      );
+
+  private final List<Bucket> sortedBuckets;
+  private final List<Interval> intervals;
+
+  public SegmentCostCache(List<Bucket> sortedBuckets)
+  {
+    this.sortedBuckets = Preconditions.checkNotNull(sortedBuckets, "buckets should not be null");
+    this.intervals = sortedBuckets.stream().map(Bucket::getInterval).collect(Collectors.toCollection(ArrayList::new));
```

Comment:
intervals supposed to be arrayList (changed field declaration from List to ArrayList)